### ISSUE TITLE
Changing the language the label does not update

### ISSUE THREAD

**Issue by matteo-gobbo** [i18next/next-i18next#2253]
## 🐛 Bug Report
Hi, when I change la language using link at the first render  the label doesn't update.

## To Reproduce

A [codesandbox](https://codesandbox.io/p/github/matteo-gobbo/next-12-with-i18next-issue/main?layout=%257B%2522sidebarPanel%2522%253A%2522GIT%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clrj2cfsx00063b6m6akn5uuc%2522%252C%2522sizes%2522%253A%255B51.77727784026997%252C48.22272215973003%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clrj2cfsx00023b6mvk564fwm%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clrj2cfsx00043b6meksxuoyq%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clrj2cfsx00053b6msav473ju%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B70.43661986822637%252C29.563380131773627%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clrj2cfsx00023b6mvk564fwm%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrj2cfsx00013b6megc5jurd%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clrj2cfsx00023b6mvk564fwm%2522%252C%2522activeTabId%2522%253A%2522clrj2cfsx00013b6megc5jurd%2522%257D%252C%2522clrj2cfsx00053b6msav473ju%2522%253A%257B%2522id%2522%253A%2522clrj2cfsx00053b6msav473ju%2522%252C%2522activeTabId%2522%253A%2522clrj2gn5600tg3b6mlbq6u85i%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clrj2gn5600tg3b6mlbq6u85i%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252Fdemo%2522%257D%255D%257D%252C%2522clrj2cfsx00043b6meksxuoyq%2522%253A%257B%2522id%2522%253A%2522clrj2cfsx00043b6meksxuoyq%2522%252C%2522activeTabId%2522%253A%2522clrj3xl0f00303b6mgwnudsc8%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clrj2cfsx00033b6mdl9tbrvq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clrj2cgan000ve6es1sqr24tu%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522CSB_RUN_OUTSIDE_CONTAINER%253D1%2520devcontainer%2520templates%2520apply%2520--template-id%2520%255C%2522ghcr.io%252Fdevcontainers%252Ftemplates%252Ftypescript-node%255C%2522%2520--template-args%2520%27%257B%257D%27%2520--features%2520%27%255B%255D%27%2522%252C%2522id%2522%253A%2522clrj2ctjk00813b6mbekw68fc%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522id%2522%253A%2522clrj2gl8n00qc3b6mphx652cg%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522GIT_LOG%2522%252C%2522shellId%2522%253A%2522clrj3xio50024e6es9bdu3yhi%2522%252C%2522id%2522%253A%2522clrj3xl0f00303b6mgwnudsc8%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) example 
To reproduce:
- Go to /demo
- Click on EN
- You will see at first render the old label

## Expected behavior

When changing the path I expect that the label is updated too. 

## Your Environment

- *runtime*: node v18.17.0
- *i18next*: ^23.7.16
- *next-i18next*: ^15.0.0
- *react-i18next*: ^13.5.0
- *os*: Mac

**Comment by adrai** [comment]
That's probably because you set the locale to 'it' here:

```ts
export const getStaticProps: GetStaticProps = async () => {
    const locale = "it";

    return {
        props: {
            locale,
            ...(await serverSideTranslations(locale, ["common"])),
        },
    };
};
```

do this instead:

```ts
export const getStaticProps: GetStaticProps<Props> = async ({
  locale,
}) => ({
  props: {
    locale,
    ...(await serverSideTranslations(locale ?? 'it', [
      'common'
    ])),
  },
})
```

Why do you have an "en" folder?
Please compare with this example: https://github.com/i18next/next-i18next/tree/master/examples/simple
Or if you use ssg with this one: https://github.com/i18next/next-i18next/tree/master/examples/ssg

**Comment by matteo-gobbo** [comment]
Thank you @adrai for your reply.

Sadly this is only a replica of a bigger project that is structured this way with all italian pages on one side and all english pages under the folder `en`. So this is way you see locale hardcoded to "it" or "en".

**Comment by alfredosalzillo** [comment]
I'm experiencing the same problem, using the 14.* version solve the issue.

**Comment by matteo-gobbo** [comment]
Yes @alfredosalzillo we're actually using the v14 and it works fine. While trying to upgrade to v15 I found that problem.

**Comment by rommmy** [comment]
Hi there, 

Currently experiencing the same issue while trying to upgrade from v14 to v15.

My app only uses serverSideTranslations. The behaviour is the same as the codesandbox linked by @matteo-gobbo: Loading the localized homepage, switching language (through a <a> link pointing to the localized homepage), and then the translations don't get updated until I manually refresh the page.

After some debugging, I found out that if I replace the use of useLayoutEffect with a useEffect, everything works fine. I don't know if it lets the door open to some other kind of problems, but at least using useEffect I can switch the language and see the correct translations applied without having to refresh the page.

My first guess would be that the call to i18n.changeLanguage inside useLayoutEffect causes a re-render that would lead React to update the DOM in an unexpected order. Might also be something else.

Any thoughts?
